### DISCUSSION TITLE
Dropbox context menu item => Copy Dropbox link

### DISCUSSION THREAD

**Discussion by ghost** [j2969719/doublecmd-plugins#32] [General]
Is there any plugin or action or whatever that could work in conjunction with the official desktop app in Ubuntu/Mint to provide a simple context menu item that lets the user copy the Dropbox link to the selected file/folder? That's all I need and apparently the Cloud plugin by @ivanenko is way too obscure in regard to its usage.

There is a Cinnamon -> Nemo plugin that does that and it's the only thing that makes me still use it when necessary. I'd like to eliminate that "dependency".

**Comment by j2969719** [comment]
double commander uses the same plugin api as total commander, unfortunately it doesnt provide the conventional ability to touch things like the context menu.

but you can add additional actions to the context menu using internal file associations.

keep in mind that due to lack of free time, cant check whether the following will work

a quick google search turned up [this page](https://help.dropbox.com/ru-ru/installs/linux-commands) with available dropbox commands so eg:
in DC in `options->file associations` add some new file type
in the `extensions` field add reserved extension `file` (or `folder`)
in the `actions` field add some custom action
in the `command` field enter `{!SHELL}`
in the `parameters` field copy from the dropbox documentation, for example, the same `dropbox puburl FILE`, replacing `FILE` with `%p`

more info:
https://doublecmd.github.io/doc/en/configuration.html#ConfigAssociations
https://doublecmd.github.io/doc/en/variables.html

**Reply by ghost** [reply]
Thanks for trying to help, however I tried all possible combinations I could think of and haven't gotten the desired result.
Just to make it clear about the final result: the Dropbox URL should be copied to clipboard upon selecting the 'Copy Dropbox link' action.
So far the closest I could get to the desired result was a terminal window popup with the URL, waiting for pressing ENTER to close. That is not optimal at all.

Maybe I didn't understand the usage of `Get output from command` option...? Is there a way to feed the output of the dropbox command directly to the clipboard?

The correct command for my case would be `dropbox sharelink %p` which should yield a URL of the form:
`https://www.dropbox.com/s/<random_alnum_string>/<simple_file_name>?dl=0`

**EDIT:**
OK, after a little bit more fiddling got a step closer to a usable result but still not perfect. This is to be placed in the **Parameters** field:
`dropbox sharelink %p | xsel -b && xdotool sleep 1 && xdotool key Return`
and the **Start path** field should contain `%D`.

**xsel** is not installed by default in certain Linux distros (such as Mint 19) so the user may have to install it manually.
First I tried with **xclip** but that one has the nasty habit of clearing the clipboard when terminal window is closed, so not good.
The intermediary **sleep** is needed for a lazy terminal window just to make sure the Return command goes to it and not to some other window.

The only inconvenience left is an extra line feed in the end of the copied string. Fixing that would require too much additional fuss so will leave it as is.

**Comment by j2969719** [comment]
>Maybe I didn't understand the usage of Get output from command option...? 

if you enter `{!VIEWER}`/`{!EDITOR}` or `mousepad`/`gedit` in the command field, and say `<?uname -a?>` in the options line, then a temporary file will be opened with `uname -a` output
>Is there a way to feed the output of the dropbox command directly to the clipboard?

with DC only , you can access the clipboard using [lua](https://doublecmd.github.io/doc/en/lua.html), but this method will work for additional buttons on the toolbar, but not in the context menu

if you only need to change the clipboard, then you don't need a terminal
create a script in python or bash eg
```bash
#!/bin/bash

file=$1
link=`dropbox sharelink "$file"`

if [ -n "$link" ]; then
  echo "$link" | xsel -b
  notify-send -i edit-copy Dropbox "$link"
else
  notify-send -i edit-copy Dropbox "Failed to get link."
fi 
```
and in the **command** field, specify the full path to the script (the first button from the input field), and in the **parameters** field, enter only `%p`

**start path** can be left blank, unless you need to forcefully specify a directory

**Reply by ghost** [reply]
> [...] a temporary file will be opened [...]

Ah, got it; not useful for this particular case but good to know for the future.

> if you only need to change the clipboard, then you don't need a terminal [...] **start path** can be left blank, unless you need to forcefully specify a directory

Unless I am mistaken the `dropbox` command **must be** launched from the system Dropbox folder, so setting the start path appears to be mandatory. For that reason a Bash or Python script may become more complicated than necessary by at least requiring a directory change, unless the script is placed precisely in the Dropbox folder, which however would make it integral part of the shared files.

To be honest I would have expected a built-in helper function that copied the result of the command to the clipboard automatically, something similar to `Get output from command`. Unfortunately, even if such command would be implemented in a future version of DC I couldn't benefit from it seeing the Bionic builds have stopped at v1.0.6 and I can't/won't upgrade this Mint 19.2 to anything higher. Oh well...

Thank you for all the help!
